File: backend/src/main/java/com/currencyexchange/service/ExternalRateService.java
package com.currencyexchange.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;

/**
 * 外部汇率API服务
 * 功能1: Get Source Currency Rates - 获取实时汇率数据
 * 
 * 数据来源: UniRateAPI (https://unirateapi.com)
 * - 免费无限制使用
 * - 支持593种货币
 * - 实时更新
 * - 26年历史数据
 */
@Service
@Slf4j
public class ExternalRateService {

    private final WebClient webClient;
    private final ObjectMapper objectMapper;
    
    @Value("${currency-exchange.api.external-api.base-url:https://api.exchangerate-api.com/v4/latest}")
    private String baseUrl;
    
    @Value("${currency-exchange.api.external-api.api-key:}")
    private String apiKey;

    public ExternalRateService() {
        this.webClient = WebClient.builder()
                .codecs(configurer -> configurer.defaultCodecs().maxInMemorySize(1024 * 1024))
                .build();
        this.objectMapper = new ObjectMapper();
    }

    /**
     * 获取所有当前汇率 (基于USD)
     * GET /v4/latest/USD (exchangerate-api.com格式)
     */
    public Map<String, Object> getAllCurrentRates() {
        try {
            // 使用exchangerate-api.com的正确格式
            String url = baseUrl + "/USD";
            log.info("调用外部汇率API: {}", url);
            log.info("Base URL配置: {}", baseUrl);
            
            String response = webClient.get()
                    .uri(url)
                    .retrieve()
                    .bodyToMono(String.class)
                    .block();

            log.info("外部API响应: {}", response);
            return parseExchangeRateApiResponse(response, "USD");
            
        } catch (Exception e) {
            log.error("外部汇率API调用失败: {}", e.getMessage(), e);
            return getFallbackRates();
        }
    }

    /**
     * 获取特定货币对的汇率
     * @param fromCurrency 源货币
     * @param toCurrency 目标货币
     */
    public BigDecimal getSpecificRate(String fromCurrency, String toCurrency) {
        try {
            if (fromCurrency.equals(toCurrency)) {
                return BigDecimal.ONE;
            }

            String url = buildUrl("/api/rates", fromCurrency);
            
            String response = webClient.get()
                    .uri(url)
                    .retrieve()
                    .bodyToMono(String.class)
                    .block();

            JsonNode rootNode = objectMapper.readTree(response);
            JsonNode ratesNode = rootNode.get("rates");
            
            if (ratesNode != null && ratesNode.has(toCurrency)) {
                return new BigDecimal(ratesNode.get(toCurrency).asText());
            }
            
            return getFallbackRate(fromCurrency, toCurrency);
            
        } catch (Exception e) {
            log.error("Error fetching specific rate {}/{}: {}", fromCurrency, toCurrency, e.getMessage());
            return getFallbackRate(fromCurrency, toCurrency);
        }
    }

    /**
     * 获取历史汇率
     * @param date 日期 (YYYY-MM-DD)
     * @param fromCurrency 源货币
     */
    public Map<String, Object> getHistoricalRates(String date, String fromCurrency) {
        try {
            String url = buildHistoricalUrl("/api/historical/rates", date, fromCurrency);
            
            String response = webClient.get()
                    .uri(url)
                    .retrieve()
                    .bodyToMono(String.class)
                    .block();

            return parseHistoricalResponse(response, date, fromCurrency);
            
        } catch (Exception e) {
            log.error("Error fetching historical rates for {}: {}", date, e.getMessage());
            return getFallbackHistoricalRates(date, fromCurrency);
        }
    }

    /**
     * 转换货币
     * @param amount 金额
     * @param fromCurrency 源货币
     * @param toCurrency 目标货币
     */
    public Map<String, Object> convertCurrency(BigDecimal amount, String fromCurrency, String toCurrency) {
        try {
            String url = buildConvertUrl("/api/convert", amount, fromCurrency, toCurrency);
            
            String response = webClient.get()
                    .uri(url)
                    .retrieve()
                    .bodyToMono(String.class)
                    .block();

            return parseConvertResponse(response);
            
        } catch (Exception e) {
            log.error("Error converting currency: {}", e.getMessage());
            return getFallbackConversion(amount, fromCurrency, toCurrency);
        }
    }

    /**
     * 构建API URL
     */
    private String buildUrl(String endpoint, String baseCurrency) {
        StringBuilder url = new StringBuilder(baseUrl + endpoint);
        url.append("?from=").append(baseCurrency);
        
        if (apiKey != null && !apiKey.isEmpty()) {
            url.append("&api_key=").append(apiKey);
        }
        
        return url.toString();
    }

    private String buildHistoricalUrl(String endpoint, String date, String baseCurrency) {
        StringBuilder url = new StringBuilder(baseUrl + endpoint);
        url.append("?date=").append(date);
        url.append("&from=").append(baseCurrency);
        
        if (apiKey != null && !apiKey.isEmpty()) {
            url.append("&api_key=").append(apiKey);
        }
        
        return url.toString();
    }

    private String buildConvertUrl(String endpoint, BigDecimal amount, String from, String to) {
        StringBuilder url = new StringBuilder(baseUrl + endpoint);
        url.append("?amount=").append(amount);
        url.append("&from=").append(from);
        url.append("&to=").append(to);
        
        if (apiKey != null && !apiKey.isEmpty()) {
            url.append("&api_key=").append(apiKey);
        }
        
        return url.toString();
    }

    /**
     * 解析汇率响应 (原UniRateAPI格式)
     */
    private Map<String, Object> parseRatesResponse(String response, String baseCurrency) throws Exception {
        JsonNode rootNode = objectMapper.readTree(response);
        
        Map<String, Object> result = new HashMap<>();
        result.put("base", baseCurrency);
        result.put("last_updated", LocalDateTime.now());
        result.put("source", "UniRateAPI");
        
        Map<String, BigDecimal> rates = new HashMap<>();
        JsonNode ratesNode = rootNode.get("rates");
        
        if (ratesNode != null) {
            ratesNode.fields().forEachRemaining(entry -> {
                rates.put(entry.getKey(), new BigDecimal(entry.getValue().asText()));
            });
        }
        
        result.put("rates", rates);
        return result;
    }

    /**
     * 解析ExchangeRate-API响应格式
     */
    private Map<String, Object> parseExchangeRateApiResponse(String response, String baseCurrency) throws Exception {
        JsonNode rootNode = objectMapper.readTree(response);
        
        Map<String, Object> result = new HashMap<>();
        result.put("base", baseCurrency);
        result.put("last_updated", LocalDateTime.now());
        result.put("source", "ExchangeRate-API");
        
        Map<String, BigDecimal> rates = new HashMap<>();
        JsonNode ratesNode = rootNode.get("rates");
        
        if (ratesNode != null) {
            ratesNode.fields().forEachRemaining(entry -> {
                rates.put(entry.getKey(), new BigDecimal(entry.getValue().asText()));
            });
        }
        
        result.put("rates", rates);
        log.info("成功解析汇率数据，包含{}种货币", rates.size());
        return result;
    }

    private Map<String, Object> parseHistoricalResponse(String response, String date, String baseCurrency) throws Exception {
        JsonNode rootNode = objectMapper.readTree(response);
        
        Map<String, Object> result = new HashMap<>();
        result.put("date", date);
        result.put("base", baseCurrency);
        result.put("source", "UniRateAPI");
        
        Map<String, BigDecimal> rates = new HashMap<>();
        JsonNode ratesNode = rootNode.get("rates");
        
        if (ratesNode != null) {
            ratesNode.fields().forEachRemaining(entry -> {
                rates.put(entry.getKey(), new BigDecimal(entry.getValue().asText()));
            });
        }
        
        result.put("rates", rates);
        return result;
    }

    private Map<String, Object> parseConvertResponse(String response) throws Exception {
        JsonNode rootNode = objectMapper.readTree(response);
        
        Map<String, Object> result = new HashMap<>();
        result.put("amount", rootNode.get("amount").asDouble());
        result.put("from", rootNode.get("from").asText());
        result.put("to", rootNode.get("to").asText());
        result.put("result", rootNode.get("result").asDouble());
        result.put("calculated_at", LocalDateTime.now());
        
        return result;
    }

    /**
     * 备用汇率数据 (当外部API不可用时)
     * 更新为2025年7月当前实际汇率
     */
    private Map<String, Object> getFallbackRates() {
        Map<String, Object> fallback = new HashMap<>();
        fallback.put("base", "USD");
        fallback.put("last_updated", LocalDateTime.now());
        fallback.put("source", "Fallback Data (Updated 2025-07)");
        
        Map<String, BigDecimal> rates = new HashMap<>();
        // 更新为当前实际汇率 (2025年7月)
        rates.put("EUR", new BigDecimal("0.9100"));   // 欧元较强
        rates.put("GBP", new BigDecimal("0.7800"));   // 英镑
        rates.put("JPY", new BigDecimal("155.0000")); // 日元较弱
        rates.put("CNY", new BigDecimal("7.2500"));   // 人民币当前汇率
        rates.put("KRW", new BigDecimal("1340.0000"));// 韩元较弱
        rates.put("MXN", new BigDecimal("18.5000"));  // 墨西哥比索
        rates.put("CAD", new BigDecimal("1.3600"));   // 加元
        rates.put("AUD", new BigDecimal("1.4900"));   // 澳元
        rates.put("CHF", new BigDecimal("0.8900"));   // 瑞士法郎
        rates.put("SGD", new BigDecimal("1.3500"));   // 新加坡元
        
        fallback.put("rates", rates);
        log.info("使用更新的备用汇率数据 (2025-07)");
        return fallback;
    }

    private BigDecimal getFallbackRate(String fromCurrency, String toCurrency) {
        // 简单的备用汇率计算
        Map<String, BigDecimal> usdRates = Map.of(
            "EUR", new BigDecimal("0.8500"),
            "GBP", new BigDecimal("0.7500"),
            "JPY", new BigDecimal("110.0000"),
            "CNY", new BigDecimal("6.4500"),
            "KRW", new BigDecimal("1180.0000"),
            "MXN", new BigDecimal("17.9900")
        );
        
        if (fromCurrency.equals("USD") && usdRates.containsKey(toCurrency)) {
            return usdRates.get(toCurrency);
        } else if (toCurrency.equals("USD") && usdRates.containsKey(fromCurrency)) {
            return BigDecimal.ONE.divide(usdRates.get(fromCurrency), 6, BigDecimal.ROUND_HALF_UP);
        }
        
        return BigDecimal.ONE; // 默认返回1:1
    }

    private Map<String, Object> getFallbackHistoricalRates(String date, String baseCurrency) {
        Map<String, Object> fallback = new HashMap<>();
        fallback.put("date", date);
        fallback.put("base", baseCurrency);
        fallback.put("source", "Fallback Historical Data");
        fallback.put("rates", getFallbackRates().get("rates"));
        
        return fallback;
    }

    private Map<String, Object> getFallbackConversion(BigDecimal amount, String fromCurrency, String toCurrency) {
        BigDecimal rate = getFallbackRate(fromCurrency, toCurrency);
        BigDecimal result = amount.multiply(rate);
        
        Map<String, Object> conversion = new HashMap<>();
        conversion.put("amount", amount);
        conversion.put("from", fromCurrency);
        conversion.put("to", toCurrency);
        conversion.put("result", result);
        conversion.put("calculated_at", LocalDateTime.now());
        
        return conversion;
    }
} 

File: backend2/services/ExternalRateService.js
const axios = require('axios');
const moment = require('moment');

/**
 * 外部汇率API服务
 * 功能1: Get Source Currency Rates - 获取实时汇率数据
 * 
 * 数据来源: ExchangeRate-API (https://api.exchangerate-api.com)
 * - 免费使用限制
 * - 支持多种货币
 * - 实时更新
 */
class ExternalRateService {
  constructor() {
    this.apiKey = '4919223a3cdb66f8fbf9a144';
    this.baseUrl = 'https://v6.exchangerate-api.com/v6/'+  this.apiKey +'/latest/USD';
    this.timeout = parseInt(process.env.EXTERNAL_API_TIMEOUT) || 10000;
    
    this.axiosInstance = axios.create({
      timeout: this.timeout,
      headers: {
        'User-Agent': 'Currency-Exchange-NodeJS/1.0.0'
      }
    });
  }

  /**
   * 获取所有当前汇率 (基于USD)
   * GET /v6/latest/USD (exchangerate-api.com格式)
   */
  async getAllCurrentRates() {
    try {
      const url = this.baseUrl;
      console.log('调用外部汇率API:', url);
      console.log('Base URL配置:', this.baseUrl);
      
      const response = await this.axiosInstance.get(url);
      console.log('外部API响应:', JSON.stringify(response.data, null, 2));
      
      return this.parseExchangeRateApiResponse(response.data, 'USD');
      
    } catch (error) {
      console.error('外部汇率API调用失败:', error.message);
      return this.getFallbackRates();
    }
  }

  /**
   * 获取特定货币对的汇率
   * @param {string} fromCurrency 源货币
   * @param {string} toCurrency 目标货币
   */
  async getSpecificRate(fromCurrency, toCurrency) {
    try {
      if (fromCurrency === toCurrency) {
        return 1.0;
      }

      const url = `${this.baseUrl}/${fromCurrency}`;
      
      const response = await this.axiosInstance.get(url);
      const rates = response.data.rates;
      
      if (rates && rates[toCurrency]) {
        return parseFloat(rates[toCurrency]);
      }
      
      return this.getFallbackRate(fromCurrency, toCurrency);
      
    } catch (error) {
      console.error(`Error fetching specific rate ${fromCurrency}/${toCurrency}:`, error.message);
      return this.getFallbackRate(fromCurrency, toCurrency);
    }
  }

  /**
   * 获取历史汇率
   * @param {string} date 日期 (YYYY-MM-DD)
   * @param {string} fromCurrency 源货币
   */
  async getHistoricalRates(date, fromCurrency) {
    try {
      // ExchangeRate-API 的免费版本可能不支持历史数据
      // 这里返回模拟的历史数据
      console.warn(`Historical rates not supported in free API, returning fallback for ${date}`);
      return this.getFallbackHistoricalRates(date, fromCurrency);
      
    } catch (error) {
      console.error(`Error fetching historical rates for ${date}:`, error.message);
      return this.getFallbackHistoricalRates(date, fromCurrency);
    }
  }

  /**
   * 转换货币
   * @param {number} amount 金额
   * @param {string} fromCurrency 源货币
   * @param {string} toCurrency 目标货币
   */
  async convertCurrency(amount, fromCurrency, toCurrency) {
    try {
      const rate = await this.getSpecificRate(fromCurrency, toCurrency);
      const result = amount * rate;
      
      return {
        amount: amount,
        from: fromCurrency,
        to: toCurrency,
        result: result,
        rate: rate,
        calculated_at: new Date().toISOString()
      };
      
    } catch (error) {
      console.error('Error converting currency:', error.message);
      return this.getFallbackConversion(amount, fromCurrency, toCurrency);
    }
  }

  /**
   * 解析ExchangeRate-API响应格式
   */
  parseExchangeRateApiResponse(responseData, baseCurrency) {
    const result = {
      base: baseCurrency,
      last_updated: new Date().toISOString(),
      source: 'ExchangeRate-API'
    };

    const rates = {};
    const ratesData = responseData.rates;

    if (ratesData) {
      Object.keys(ratesData).forEach(currency => {
        rates[currency] = parseFloat(ratesData[currency]);
      });
    }

    result.rates = rates;
    console.log(`成功解析汇率数据，包含${Object.keys(rates).length}种货币`);
    return result;
  }

  /**
   * 备用汇率数据 (当外部API不可用时)
   * 更新为2025年7月当前实际汇率
   */
  getFallbackRates() {
    const fallback = {
      base: 'USD',
      last_updated: new Date().toISOString(),
      source: 'Fallback Data (Updated 2025-07)'
    };

    const rates = {
      // 更新为当前实际汇率 (2025年7月)
      'EUR': 0.9100,   // 欧元较强
      'GBP': 0.7800,   // 英镑
      'JPY': 155.0000, // 日元较弱
      'CNY': 7.2500,   // 人民币当前汇率
      'KRW': 1340.0000,// 韩元较弱
      'MXN': 18.5000,  // 墨西哥比索
      'CAD': 1.3600,   // 加元
      'AUD': 1.4900,   // 澳元
      'CHF': 0.8900,   // 瑞士法郎
      'SGD': 1.3500,   // 新加坡元
      'HKD': 7.8000,   // 港币
      'INR': 83.2000,  // 印度卢比
      'BRL': 5.4000,   // 巴西雷亚尔
      'RUB': 88.0000,  // 俄罗斯卢布
      'PLN': 4.0000,   // 波兰兹罗提
      'TRY': 34.0000,  // 土耳其里拉
      'NOK': 10.8000,  // 挪威克朗
      'SEK': 10.9000,  // 瑞典克朗
      'DKK': 6.8000    // 丹麦克朗
    };

    fallback.rates = rates;
    console.log('使用更新的备用汇率数据 (2025-07)');
    return fallback;
  }

  getFallbackRate(fromCurrency, toCurrency) {
    // 简单的备用汇率计算
    const usdRates = {
      'EUR': 0.9100,
      'GBP': 0.7800,
      'JPY': 155.0000,
      'CNY': 7.2500,
      'KRW': 1340.0000,
      'MXN': 18.5000,
      'CAD': 1.3600,
      'AUD': 1.4900,
      'CHF': 0.8900,
      'SGD': 1.3500
    };
    
    if (fromCurrency === 'USD' && usdRates[toCurrency]) {
      return usdRates[toCurrency];
    } else if (toCurrency === 'USD' && usdRates[fromCurrency]) {
      return 1 / usdRates[fromCurrency];
    } else if (usdRates[fromCurrency] && usdRates[toCurrency]) {
      // 通过USD转换
      return usdRates[toCurrency] / usdRates[fromCurrency];
    }
    
    return 1.0; // 默认返回1:1
  }

  getFallbackHistoricalRates(date, baseCurrency) {
    const fallback = {
      date: date,
      base: baseCurrency,
      source: 'Fallback Historical Data',
      rates: this.getFallbackRates().rates
    };
    
    return fallback;
  }

  getFallbackConversion(amount, fromCurrency, toCurrency) {
    const rate = this.getFallbackRate(fromCurrency, toCurrency);
    const result = amount * rate;
    
    return {
      amount: amount,
      from: fromCurrency,
      to: toCurrency,
      result: result,
      rate: rate,
      calculated_at: new Date().toISOString()
    };
  }
}

module.exports = new ExternalRateService(); 